app/src/main/java/com/eveningoutpost/dexdrip/Models/Bubble.java

```diff
@@ -0,0 +1,290 @@
+    }
+
+    public static int lens = 344;
+
+    public static String serNumber(String uid) {
+        StringBuilder temp2 = new StringBuilder();
+        for (int i = uid.length(); i > 0; i = i - 2) {
+            String temp = uid.substring(i - 2, i);
+//            System.out.println(temp);
+//            System.out.println(i);
+            temp2.append(temp);
+        }
+        uid = temp2.toString();
+        StringBuilder sb = new StringBuilder();
+        try {
+            char snParts[] = {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'C', 'D', 'E', 'F',
+                    'G', 'H', 'J', 'K', 'L', 'M', 'N', 'P', 'Q', 'R', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z'};
+            String uidStr = uid.substring(4, uid.length());
+            long number = Long.parseLong(uidStr, 16);
+            String s = Long.toBinaryString(number) + "00";
+//            System.out.println(s);
+
+            for (int i = 0; i < s.length(); i = i + 5) {
+//                System.out.println(i);
+                String temp = s.substring(i, i + 5);
+                int index = toInt(temp);
+                sb.append(snParts[index]);
+
+            }
+        } catch (Exception e) {
+            e.printStackTrace();
+            return "";
+        }
+        return "0" + sb.toString();
+    }
+
+    public static int toInt(String bi) {
+        int len = bi.length();
+        int sum = 0;
+        int tmp, max = len - 1;
+        for (int i = 0; i < len; ++i) {
+            tmp = bi.charAt(i) - '0';
+            sum += tmp * Math.pow(2, max--);
+        }
+        return sum;
+    }
+
+    static int errorCount = 0;
+
+    public static BridgeResponse decodeBubblePacket(byte[] buffer, int len) {
+        final BridgeResponse reply = new BridgeResponse();
+        int count = Pref.getInt("Cobblecount", 0);
+//        Log.e(TAG, "decodeBubblePacket" + Arrays.toString(buffer));
+        if (buffer[0] == -128) {
+            PersistentStore.setString("Bubblebattery", Integer.toString(buffer[4]));
+            Pref.setInt("bridge_battery", buffer[4]);
+//            String data=DexCollectionService.bytesToHexString(buffer);
+            PersistentStore.setString("BubbleHArdware", Integer.toString(buffer[2]) + ".0");
+            PersistentStore.setString("BubbleFirmware", Integer.toString(buffer[1]) + ".0");
+//            PersistentStore.setString("LibreSN", SensorSn);
+            ByteBuffer ackMessage = ByteBuffer.allocate(6);
+            ackMessage.put(0, (byte) 0x02);
+            ackMessage.put(1, (byte) 0x01);
+            ackMessage.put(2, (byte) 0x00);
+            ackMessage.put(3, (byte) 0x00);
+            ackMessage.put(4, (byte) 0x00);
+            ackMessage.put(5, (byte) 0x2B);
+            reply.add(ackMessage);
+            bytes = null;
+            dateAll = false;
+            s_full_data = null;
+            Log.e(TAG, "reset data");
+            return getBubbleResponse();
+        }
+        if (buffer[0] == -64) {
```

Comment:
For clarity these negative integers should be their positive values but cast to a byte.